### DISCUSSION TITLE
[Show & Tell] Bulk creation of stock locations and part categories

### DISCUSSION THREAD

**Discussion by wolflu05** [inventree/InvenTree#4084] [Show and tell]
Hello,

I want to share the first beta version 🌟 of my [inventree-bulk-plugin](https://github.com/wolflu05/inventree-bulk-plugin). Currently it's fully functional, but not fully tested. I'd love if some of you find it interesting and would give me feedback.

It is a bulk creation plugin for [InvenTree](https://inventree.org/), which helps you generating locations/categories in bulk by using customized naming strategies and ensure them along your complete storage tree. That means you not only have the option to generate multidimensional* names for stock locations or part categories, but also have the option to save the templates for later usage if your storage room uses e.g. drawer towers to ensure naming consistency for later added towers.

> ℹ️ * multidimensional means that you are not limited to namings like `D1`,`D2`, .. but also something like `D1.A`, `D1.B`, `D2.A`, `D2.B`, ...

### 🌟Screenshots

<details open>
<summary>This will generate the previous mentioned example:</summary>

![image](https://user-images.githubusercontent.com/76838159/208684801-92dc9f1c-40c4-4240-8196-b292c837460f.png)

</details>

### 📄 Some more examples

<details>
<summary>But even nested generations don't stop you.</summary>

![image](https://user-images.githubusercontent.com/76838159/208685037-7e9d7ed7-d760-4e7d-addc-fcf65d5c2286.png)

</details>

<details>
<summary>And with parent name match, you can even generate different sub-locations depending of the name of the parent:</summary>

![image](https://user-images.githubusercontent.com/76838159/208685579-85fc6729-7146-4ce5-a31b-c1e7e85b4a9e.png)

</details>

### ℹ️ More info

You'll find a complete usage and installation guide on my [README.md](https://github.com/wolflu05/inventree-bulk-plugin#readme).

I already know that @joaoferreira-git was trying to build something like this with ChatGPT in #4016.

**Comment by SergeoLacruz** [comment]
Hi,
Nice Idea. I needed this a year ago... i just wrote some python code using the python lib cause I needed it only once for the first setup. I will try your plugin soon. It is also a good example. Thanks for that. 

Michael

**Reply by wolflu05** [reply]
Hey Michael,

I was at the same point and wanted something more usable for non-programmers. Remember this is currently in a beta phase (v0.1.0) and should not be used in production. I'd love to get this released as soon as it gets more tested.

Lukas

**Comment by toinopt** [comment]
For now i don't really need this but if i have the time this week, i will install this in production and test it, if it breaks i'm going to ask for a refund and start a lawsuit for all the emocional damage. 😂

**Reply by wolflu05** [reply]
Hi joaoferreira-git,

I hope that also the fact to install this in production even you don't need it was irony. After the installation you have one extra table in your database which saves your stored bulk creation schemas. (See [why does this plugin needs the app mixin?](https://github.com/wolflu05/inventree-bulk-plugin/#why-does-this-plugin-needs-the-app-mixin)). But have fun trying this out in a testing environment.

Lukas

**Reply by toinopt** [reply]
@wolflu05 
Just finished installing and testing this and it works like a charm, wish i had this when before i had to create over 160 locations by hand, the only extra feature i would like is to add the possibility to add extra location.
Going to try and explain myself the best i can.

This are the storage things i use and the last row is just a big drawer and if there was a way to create all of them with just one template it would help some people in future that might be setting up a warehouse with similar storage things.

For now i solved it by using this template `Gaveta - {dim.1}{dim.2}` to generate all of except the last one and another template with the name of the drawer `Gaveta - I1`.
What im asking for is a way to add both lines in the Generate name field something like this `Drawer - {dim.1}{dim.2}+Drawer - I1`, in my case this would generate the 32 storage locations plus the big one at the bottom named Drawer - I1.
Now i need to printscreen a bunch of qr labels, place them in a word file i have, print it, cut them one by one and then place them in the respective drawer, all of this because i don't have a label printer. 😔

Storage things i have already setup and there's a extra one waiting to be bolted into the wall.
I need to get rid of that wall light but i haven't found anything good to replace it.
![image](https://user-images.githubusercontent.com/19786598/210645640-55413aa6-a9a3-455b-ac75-91fa5d9abc08.png)
![image](https://user-images.githubusercontent.com/19786598/210645725-045e75b7-968c-4859-ba40-c79f99b276c5.png)

**Reply by SergeoLacruz** [reply]
Nice maker space. The T-shape SCHUKO divider is vintage :-) Seems you are located in Germany.

**Reply by matmair** [reply]
@SergeoLacruz not another one ;-) I would think Portugal or Spain

**Reply by wolflu05** [reply]
@joaoferreira-git Thank you for trying out my plugin. Your space looks pretty cool organized. 

For your label printing thing, maybe you can develop a standard paper printer plugin which just adds bulk prints onto a single paper until it's filled and uses the next page, so that you don't have to waste your time manually adding QR codes to a word document. 

I think your request could be handled with this feature request https://github.com/wolflu05/inventree-bulk-plugin/issues/4 . Then you could do something like `1-40,I1`,Please let me know what you think of this solution.

**Reply by matmair** [reply]
I thought we had an FR for printing labels on sheets of paper but can not find it right now. This has definitely been asked about before. Feel free to raise a feature request, not sure how this is done best (core or plugin) but this seems like something we should have a standard solution for from my POV (no promises though).

**Reply by toinopt** [reply]
> Nice maker space. The T-shape SCHUKO divider is vintage :-) Seems you are located in Germany.

I'm from Portugal and T-shape SCHUKO thing is still sold here in Portugal, this one is old but in the garage I have one I bought last year.

> @joaoferreira-git Thank you for trying out my plugin. Your space looks pretty cool organized.
> 
> For your label printing thing, maybe you can develop a standard paper printer plugin which just adds bulk prints onto a single paper until it's filled and uses the next page, so that you don't have to waste your time manually adding QR codes to a word document.
> 
> I think your request could be handled with this feature request [wolflu05/inventree-bulk-plugin#4](https://github.com/wolflu05/inventree-bulk-plugin/issues/4) . Then you could do something like `1-40,I1`,Please let me know what you think of this solution.

I think that feature would do what i need.
Trying to be organized and that's why i installed inventree, to force me to be organized and so far it's kinda working.

> I thought we had an FR for printing labels on sheets of paper but can not find it right now. This has definitely been asked about before. Feel free to raise a feature request, not sure how this is done best (core or plugin) but this seems like something we should have a standard solution for from my POV (no promises though).

No problem i can open a feature request, after i finish sorting and organizing all my stuff i will make sure to post the pictures in the a discussions as someone asked, not sure if it was you or SchrodingersGat.

**Comment by matmair** [comment]
Nice!

**Comment by SchrodingersGat** [comment]
Very cool @wolflu05 and looks great

**Reply by wolflu05** [reply]
Thanks @SchrodingersGat and @matmair. The generation on the backend is making use of pythons itertools. My next goals are to add some tests and add a few smaller features.

Currently I don't know, but I feel like the AppMixin (to store custom models) for saving stored templates) feel somehow not right, but I don't found another solution. And secondly I use the app mixin for serving preact (leightweight react which needs in combination with htm no build, it can run directly in the browser) as a static file.

And my second point is bulk creation takes some time (multiple seconds) if they are huge amounts of items (>1000) if I create each item after each other. What do you think of this solution? https://stackoverflow.com/a/67517621

**Reply by matmair** [reply]
@wolflu05 AppMixin was exactly included for these more complex use cases as it exposes all Django functionalities. And it reduces the number of things we need to expose manually.

**Reply by wolflu05** [reply]
Ok. Feel bad because this has some side effects if you remove the plugin, you still have the table on your database.

And what do you think to that Stack overflow Post where they set the lft, rgh and level to 0 and then regenerate the tree? Does this look like an appropriate solution or should I continue with the single creation?

**Reply by matmair** [reply]
@wolflu05 I do not really see a problem with persistent data storage, better than losing it. We could add a lifecycle task for removing it but that could create issues with failed migrations.

> Does this look like an appropriate solution or should I continue with the single creation?

We probably need to change the base model for that - not really sure it is worth it.

**Reply by wolflu05** [reply]
Ok. Thank you

**Comment by eeintech** [comment]
Very nice, well done!

**Reply by wolflu05** [reply]
Thank you very much, do you have ideas to improve it?

**Reply by eeintech** [reply]
I only saw the screenshot. However I tried to install your plugin on my local install and couldn't... I'm stuck at the "activate plugin" step as I don't have it in my list of plugins after server/workers restart:

![image](https://user-images.githubusercontent.com/4020546/208995954-70d1acc8-50f1-4c20-89be-8c265f7866ed.png)

@matmair I have those settings in my `config.yaml`:
```yaml
# Set this variable to True to enable InvenTree Plugins
# Alternatively, use the environment variable INVENTREE_PLUGINS_ENABLED
plugins_enabled: True
plugin_file: '/home/francois/Desktop/github/InvenTree/InvenTree/plugins/plugins.txt'
plugin_dir: '/home/francois/Desktop/github/InvenTree/InvenTree/plugins'
```

Why couldn't I install @wolflu05's plugin?

EDIT: Don't see anymore info in the doc: https://docs.inventree.org/en/latest/extend/plugins/install/#plugin-installation-file-pip

**Reply by wolflu05** [reply]
~~Interesting, can you please share the contents of your plugins.txt? And are you sure you should down everything and booted it up? I tried the exact steps as listed on the readme multiple times on a fresh docker installation.~~

**Reply by wolflu05** [reply]
Ah, it's in your list. Just enable it. The last in your not activated ones. Click on the icon on the right and then select activate on Django admin, save and continue the guide.

**Reply by eeintech** [reply]
@wolflu05 I saw the "BulkAction" plugin but can't activate, the "active" checkbox is not persistent after saving:

![image](https://user-images.githubusercontent.com/4020546/208997575-9c38a6f0-4d2e-4ecd-b662-34d545526599.png)

I still tried to restart the server/worker after that and still not activated.

I would suggest to rename it to match your repository name so nobody can miss it :smiley:

**Reply by wolflu05** [reply]
> I would suggest to rename it to match your repository name so nobody can miss it 😃

Good idea. Added to todo.

But no idea why it's not persistent. Maybe @matmair knows something? 
1. Just to get it right, you click the activated checkbox, so that it's checked, click save and then restart and then it's not loaded, because it's plugin config is not active? 
2. Do you get any errors in terminal or Notification Center?
3. which inventree version (and commit ref) are you running? My plugin requires at min 0.9.1 (not released yet)

**Reply by eeintech** [reply]
I'm running `0.9.0 stable` on  `debug=True` mode / baremetal installation.

This is the server log after clicking "Activate" and saving:
```bash
15:53:19 [Q] INFO Enqueued 68242
[21/Dec/2022 15:53:19] "POST /admin/plugin/pluginconfig/17/change/ HTTP/1.1" 302 0
[21/Dec/2022 15:53:19] "GET /admin/plugin/pluginconfig/ HTTP/1.1" 200 42712
[21/Dec/2022 15:53:19] "GET /admin/jsi18n/ HTTP/1.1" 200 3195
[21/Dec/2022 15:53:20] "GET /favicon.ico HTTP/1.1" 302 0
[21/Dec/2022 15:53:20] "GET /index/ HTTP/1.1" 200 23550
```
There is this 302 redirect (temporary redirect) popping up right after.

Here's the worker log meanwhile
```bash
15:53:19 [Q] INFO Process-1:2 processing [louisiana-whiskey-oranges-cat]
15:53:19 [Q] INFO Processed [louisiana-whiskey-oranges-cat]
```

Here's me trying to activate it (history):

![image](https://user-images.githubusercontent.com/4020546/209001143-c7be4c69-b0e1-42ba-abd6-be2daae637d1.png)

**Reply by wolflu05** [reply]
Yeah, then I guess the MIN_VERSION is the issue. See: MIN_VERSION = "0.9.1"  # due to "invoke update" doesn't run collectstatic (see #inventree/InvenTree#4077)

**Reply by eeintech** [reply]
Alright so I:
- Switched to `master` and pulled latest
- Ran `inv update`, starter server/worker and activated plugin, the "Activate" box is persistent this time
- After activation I re-ran `inv update` and restarted server/worker

Now I'm here:

![image](https://user-images.githubusercontent.com/4020546/209003563-cbd20094-9390-4edd-80a2-e8aaf68c56dd.png)

Clicking "Open in new tab" button next to `templates` and `templatebyid` gives me a blank page with `[]` displayed.
Cliking next to `bulkcreatecategory` and `bulkcreatelocation` redirects to `/index/`.
Clicking next to `parse` gives an error:
``` bash
Internal Server Error: /plugin/bulkaction/parse
Traceback (most recent call last):
  File "/home/francois/Desktop/github/InvenTree/env-inventree/lib/python3.10/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/home/francois/Desktop/github/InvenTree/env-inventree/lib/python3.10/site-packages/django/core/handlers/base.py", line 188, in _get_response
    self.check_response(response, callback)
  File "/home/francois/Desktop/github/InvenTree/env-inventree/lib/python3.10/site-packages/django/core/handlers/base.py", line 309, in check_response
    raise ValueError(
ValueError: The view inventree_bulk.BulkActionPlugin.method.__call__ didn't return an HttpResponse object. It returned None instead.
Internal Server Error: /plugin/bulkaction/parse
Traceback (most recent call last):
  File "/home/francois/Desktop/github/InvenTree/env-inventree/lib/python3.10/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/home/francois/Desktop/github/InvenTree/env-inventree/lib/python3.10/site-packages/django/core/handlers/base.py", line 188, in _get_response
    self.check_response(response, callback)
  File "/home/francois/Desktop/github/InvenTree/env-inventree/lib/python3.10/site-packages/django/core/handlers/base.py", line 309, in check_response
    raise ValueError(
ValueError: The view inventree_bulk.BulkActionPlugin.method.__call__ didn't return an HttpResponse object. It returned None instead
```

How should I proceed?

**Reply by wolflu05** [reply]
This looks very good. See usage on my readme now. You can go to any sub-location and try out the bulk creation panel.

**Reply by eeintech** [reply]
Oh I assumed it wasn't working properly. Yes I see the tab right now. Let me give it a shot and I'll report back.

**Reply by matmair** [reply]
Seems you sorted it out

**Reply by wolflu05** [reply]
> Seems you sorted it out

@matmair yeah, but instead of don't letting the user activate the plugin if the min version check fails it would be nice if there would be an error message, why there is this strange behavior.

**Reply by eeintech** [reply]
@wolflu05 How can I use the parent name in the child description field?

This doesn't work:

![image](https://user-images.githubusercontent.com/4020546/209013851-340c1b90-dd5a-4c3c-ac72-cfd955dd061b.png)

Also, I wanted to name the children: Left Space, Middle Space and Right Space.
Dimensions field doesn't let me use a full word, should that be doable?

**Reply by wolflu05** [reply]
These are two very good ideas which I haven't thought of. Would you mind creating two feature requests as issues on my repository?

**Reply by eeintech** [reply]
Done! I updated the description to check the preview:

![image](https://user-images.githubusercontent.com/4020546/209015734-c3e2c347-0cc8-4043-bcb8-368443991a5b.png)

Looks good and it works!

![image](https://user-images.githubusercontent.com/4020546/209015897-14336725-8db0-4cc6-9738-c1fd075a8d7b.png)

Nice job @wolflu05 :wink:

**Reply by wolflu05** [reply]
Thank you. 
One Tipp, you can also use a generate name without a dimension, so you can just generate a static name in each parent as sub-location.

**Reply by wolflu05** [reply]
What do you think? In the beginning, the ui may look a bit complex to understand, but my goal was to provide that set of advanced opportunities. Was it self understanding? Do you read through the documentation? How did you managed to learn how to use it?

**Reply by eeintech** [reply]
> One Tipp, you can also use a generate name without a dimension, so you can just generate a static name in each parent as sub-location.

Makes sense, I was pushing your plugin to see how far it can go :wink: 

> What do you think? In the beginning, the ui may look a bit complex to understand, but my goal was to provide that set of advanced opportunities.

Overall I wasn't familiar with plugins, @matmair did a real nice job adding this feature in InvenTree which, with a little bit of polishing, can become more admin friendly.

About your plugin, I guess I like the idea if you were to manage a warehouse for instance. For our "little" lab I could have saved myself from creating our current sublocations (we have 20), but it didn't take much as I probably spend half hour doing it manually.

> Was it self understanding? Do you read through the documentation? How did you managed to learn how to use it?

I wouldn't have understood without the screenshot you posted above honestly, pictures/video examples are much much easier to get you started than words IMO.
I did ready your README but was still confused about a few things so the screenshot helped clarify.

So I would say the first thing to show is a simple example then dive into the details.

**Reply by wolflu05** [reply]
Thank you for your feedback. Good idea with the example. I'll also try to add some short gifs. It's hard for me as a developer who developed that plugin to say how easy it will be for others to understand, but I try my best to make it as easy as possible.

**Comment by Benjamin0001** [comment]
Any news about this plugin? Can I use it and then remove it when I'm done?

**Reply by wolflu05** [reply]
Hello, sorry to say, but I'm currently very busy, so I have no time improving it. Currently the plugin works, but I have several improvements planned. My goal is to continue development in a few months.

**Reply by Benjamin0001** [reply]
Ok, thanks for the quick reply! I will try it out. I have about 12x16x4 shelf locations that I need to create for a test run. I will let you know how it goes.

**Comment by wolflu05** [comment]
Hello all, it took a long time, but now the time has come, v1.0.0 finally arrives 🚀 on 02.08.2023. There were a lot of breaking changes during the beta period to test out different things, but now with the first final version the syntax will stay compatible. You can read more about v1.0.0 on the inventree-bulk-plugin [announcement discussion](https://github.com/wolflu05/inventree-bulk-plugin/discussions/40) or on the [README.md](https://github.com/wolflu05/inventree-bulk-plugin#readme). I would really appreciate if you can share any feedback about the plugin on the linked discussion.

**Comment by wolflu05** [comment]
v1.1.0 is released now 🚀. Now even parts can be generated using templates. You can read more about the highlights on the [v1.1.0 release page](https://github.com/wolflu05/inventree-bulk-plugin/releases/tag/1.1.0)

**Reply by SchrodingersGat** [reply]
Very impressive @wolflu05
